File: hpu_ops/backend/fbgemm.cpp
/**
 * Copyright (c) 2021-2025 Intel Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "hpu_ops/fbgemm.h"

namespace habana {

LazyPermuteSparseDataCommon::LazyPermuteSparseDataCommon(
    int device_id,
    c10::ScalarType scalar_type,
    bool is1D,
    bool hasWeights)
    : OpBackend(
          device_id,
          "permute_" + (is1D ? std::string("1D") : std::string("2D")) +
              "_sparse_data_fwd_",
          scalar_type,
          hasWeights ? std::vector<int>{1, 2, 3} : std::vector<int>{1, 2},
          {},
          {},
          false) {}

LazyPermute1DSparseData::LazyPermute1DSparseData(
    int device_id,
    c10::ScalarType scalar_type,
    bool hasWeights)
    : LazyPermuteSparseDataCommon(device_id, scalar_type, true, hasWeights) {}

LazyPermute2DSparseData::LazyPermute2DSparseData(
    int device_id,
    c10::ScalarType scalar_type,
    bool hasWeights)
    : LazyPermuteSparseDataCommon(device_id, scalar_type, false, hasWeights) {}

void LazyPermute1DSparseData::AddNode(
    synapse_helpers::graph& graph,
    const at::Stack& stack) {
  LazyPermuteSparseDataCommon::AddLazyPermuteSparseDataNode(graph, stack, true);
}

void LazyPermute2DSparseData::AddNode(
    synapse_helpers::graph& graph,
    const at::Stack& stack) {
  LazyPermuteSparseDataCommon::AddLazyPermuteSparseDataNode(
      graph, stack, false);
}

LazyExpandIntoJaggedPermute::LazyExpandIntoJaggedPermute(
    int device_id,
    c10::ScalarType scalar_type)
    : OpBackend(
          device_id,
          "expand_into_jagged_permute",
          scalar_type,
          {1},
          {},
          {},
          false) {}

void LazyExpandIntoJaggedPermute::AddNode(
    synapse_helpers::graph& graph,
    const at::Stack& stack) {
  std::vector<synTensor> inputs = {syn_in(0), syn_in(1), syn_in(2)};

  auto input_offsets = stack.at(1).toTensor();

  int64_t output_size = stack.at(3).toScalar().toInt();

  std::string guid = "expand_into_jagged_permute_fwd_i32";

  std::vector<NodeAttr::NodeOutputAttr> output_attrs = {
      {{output_size}, input_offsets.scalar_type(), 0}};

  auto permuted = OpBackend::BuildNode(
      this, graph, {guid, inputs, output_attrs, nullptr, 0});

  syn_out(0) = std::move(permuted[0]);
}

using namespace std::literals;

LazyBoundsCheckIndices::LazyBoundsCheckIndices(
    int device_id,
    c10::ScalarType scalar_type)
    : OpBackend(
          device_id,
          "bounds_check_indices_fwd_",
          scalar_type,
          {},
          {0, 1, 2},
          {},
          false) {}

void LazyBoundsCheckIndices::AddNode(
    synapse_helpers::graph& graph,
    const at::Stack& stack) {
  StackGetter stackGetter(this, stack, "LazyBoundsCheckIndices::AddNode");
  auto indices = stackGetter.getNextInput<TensorsPair>();
  auto offsets = stackGetter.getNextInput<TensorsPair>();
  auto warning = stackGetter.getNextInput<TensorsPair>();
  auto rowsPerTable = stackGetter.getNextInput<TensorsPair>();
  auto boundsCheckMode = stackGetter.getNextInput<int>();
  auto weights = stackGetter.getNextInput<std::optional<TensorsPair>>();

  std::vector<synTensor> inputs = {
      rowsPerTable.syn_t, indices.syn_t, offsets.syn_t, warning.syn_t};
  if (weights) {
    inputs.push_back(weights.value().syn_t);
  }

  std::string guid =
      get_guid_with_precision("bounds_check_indices_fwd"sv, ScalarType());

  std::vector<NodeAttr::NodeOutputAttr> output_attrs = {
      {indices.pt_t.sizes(), indices.pt_t.scalar_type(), 0},
      {offsets.pt_t.sizes(), offsets.pt_t.scalar_type(), 1},
      {warning.pt_t.sizes(), warning.pt_t.scalar_type(), 2}};

  ns_BoundsCheckIndicesKernel::Params params{};
  params.mode = static_cast<BoundsCheckMode_t>(boundsCheckMode);

  auto results = OpBackend::BuildNode(
      this, graph, {guid, inputs, output_attrs, &params, sizeof(params)});

  for (size_t i = 0; i < output_attrs.size(); ++i) {
    syn_out(i) = std::move(results[i]);
  }
}

LazySplitPermuteCat::LazySplitPermuteCat(
    int device_id,
    c10::ScalarType scalar_type)
    : OpBackend(
          device_id,
          "split_permute_cat",
          scalar_type,
          {0},
          {},
          {},
          false) {}

void LazySplitPermuteCat::AddNode(
    synapse_helpers::graph& graph,
    const at::Stack& stack) {
  StackGetter stackGetter(this, stack, "LazySplitPermuteCat::AddNode");
  auto input = stackGetter.getNextInput<TensorsPair>();
  auto indices = stackGetter.getNextInput<TensorsPair>();

  auto batchSize = stackGetter.getNextInput<int>();
  auto numFeatures = stackGetter.getNextInput<int>();
  auto dims = stackGetter.getNextInput<int>();

  std::string guid = get_guid_with_precision(
      "split_permute_cat_fwd"sv, input.pt_t.scalar_type());

  ns_SplitPermuteCat::Params params;
  params.batchSize = batchSize;
  params.numFeatures = numFeatures;
  params.dims = dims;

  auto output = OpBackend::BuildNode(
      this,
      graph,
      {guid,
       {input.syn_t, indices.syn_t},
       {{input.pt_t.sizes(), input.pt_t.scalar_type(), 0}},
       &params,
       sizeof(params)});

  syn_out(0) = std::move(output[0]);
}

} // namespace habana

static const auto& FBGEMMKernelsKernelRegistry =
    habana::KernelRegistry()
        .add(
            "hpu::habana_permute_1D_sparse_data",
            KERNEL_FN_ARG(LazyPermute1DSparseData, true))
        .add(
            "hpu::habana_permute_1D_sparse_data_without_weights",
            KERNEL_FN_ARG(LazyPermute1DSparseData, false))
        .add(
            "hpu::habana_permute_2D_sparse_data",
            KERNEL_FN_ARG(LazyPermute2DSparseData, true))
        .add(
            "hpu::habana_permute_2D_sparse_data_without_weights",
            KERNEL_FN_ARG(LazyPermute2DSparseData, false))
        .add(
            "hpu::habana_expand_into_jagged_permute",
            KERNEL_FN_GLOBAL(habana::LazyExpandIntoJaggedPermute))
        .add(
            "hpu::habana_bounds_check_indices",
            KERNEL_FN_GLOBAL(habana::LazyBoundsCheckIndices))
        .add(
            "hpu::habana_split_permute_cat",
            KERNEL_FN_GLOBAL(habana::LazySplitPermuteCat));

File: python_packages/habana_frameworks/torch/hpex/kernels/fbgemm.py
###############################################################################
#
#  Copyright (c) 2021-2025 Intel Corporation
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
#
###############################################################################


from habana_frameworks.torch import _hpex_C

import torch

# The file implements operators included in the FBGEMM (Facebook GEneral Matrix Multiplication) library.


def permute_1D_sparse_data(
    permute: torch.Tensor, lengths: torch.Tensor, indices: torch.Tensor, weights: torch.Tensor | None
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    list_res = _hpex_C.permute_1D_sparse_data(permute, lengths, indices, weights)
    return tuple(list_res) if len(list_res) == 3 else (list_res[0], list_res[1], None)


def permute_2D_sparse_data(
    permute: torch.Tensor, lengths: torch.Tensor, indices: torch.Tensor, weights: torch.Tensor | None
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
    list_res = _hpex_C.permute_2D_sparse_data(permute, lengths, indices, weights)
    return tuple(list_res) if len(list_res) == 3 else (list_res[0], list_res[1], None)


def split_embedding_codegen_lookup_function(
    host_weights: torch.Tensor,
    weights_offsets: list[int],
    D_offsets: list[int],
    total_D: int,
    indices: torch.Tensor,
    offsets: torch.Tensor,
    pooling_mode: int,
    kernel_mode: list[int] = None,
) -> torch.Tensor:
    if kernel_mode is None:
        kernel_mode = [1] * len(weights_offsets)
    assert pooling_mode == 0, "Only PoolingMode.SUM is supported for HPU"
    assert total_D == D_offsets[-1], f"total_D ({total_D}) must match D_offsets[-1] ({D_offsets[-1]})"

    indices = indices.to(torch.int32)
    offsets = offsets.to(torch.int32)

    T = len(D_offsets) - 1
    B = (offsets.size(dim=0) - 1) // T

    outputs = []

    previous_D = D_offsets[1]
    for t in range(T):
        D = D_offsets[t + 1] - D_offsets[t]

        assert previous_D == D, f"HPU supports only constant D_offsets' distances, but they're {D} and {previous_D}"

        t_weights_from = weights_offsets[t]
        if t + 1 < T:
            t_weights_to = weights_offsets[t + 1]
        else:
            t_weights_to = host_weights.size(dim=0)

        # Since t_weights_from and t_weights_to are int64, we need to do
        # reshape before slice, not to pass int64 to slice kernel. This also
        # forces an assumption that D should be constant
        t_weights = host_weights.reshape(
            host_weights.size(dim=0) // D, D  # Number of rows (words)  # number of cols (words' meanings)
        )[t_weights_from // D : t_weights_to // D]

        t_offsets = offsets[t * B : (t + 1) * B + 1]
        t_indices = indices[t * B : (t + 1) * B]

        if kernel_mode[t] == 0:
            valid_count = torch.tensor([t_indices.numel()], dtype=torch.int32, device="hpu")
        elif kernel_mode[t] == 2:
            valid_count = torch.tensor([t_offsets.numel()], dtype=torch.int32, device="hpu")
        else:
            valid_count = torch.tensor([t_offsets.numel(), t_offsets.numel()], dtype=torch.int32, device="hpu")
        emb_out = _hpex_C.embedding_bag_sum_fwd(
            t_weights, indices if kernel_mode[t] else t_indices, t_offsets, valid_count, kernel_mode[t]
        )

        outputs.append(emb_out)

    return torch.cat(outputs, dim=1)


def split_embedding_codegen_lookup_sgd_function_hpu(
    host_weights: torch.Tensor,
    weights_placements: torch.Tensor,
    weights_offsets: list[int],
    D_offsets: list[int],
    total_D: int,
    max_D: int,
    hash_size_cumsum: torch.Tensor,
    total_hash_size_bits: int,
    indices: torch.Tensor,
    offsets: torch.Tensor,
    pooling_mode: int,
    indice_weights: torch.Tensor | None,
    feature_requires_grad: torch.Tensor | None,
    gradient_clipping: bool,
    max_gradient: float,
    stochastic_rounding: bool,
    learning_rate: int = 0,
    output_dtype: int = 0,
) -> torch.Tensor:
    return split_embedding_codegen_lookup_function(
        host_weights, weights_offsets, D_offsets, total_D, indices, offsets, pooling_mode
    )


def split_embedding_codegen_lookup_adagrad_function_hpu(
    host_weights: torch.Tensor,
    weights_placements: torch.Tensor,
    weights_offsets: list[int],
    D_offsets: list[int],
    total_D: int,
    max_D: int,
    hash_size_cumsum: torch.Tensor,
    total_hash_size_bits: int,
    indices: torch.Tensor,
    offsets: torch.Tensor,
    pooling_mode: int,
    indice_weights: torch.Tensor | None,
    feature_requires_grad: torch.Tensor | None,
    gradient_clipping: bool,
    max_gradient: float,
    stochastic_rounding: bool,
    momentum1_host: torch.Tensor,
    momentum1_placements: torch.Tensor,
    momentum1_offsets: torch.Tensor,
    learning_rate: int = 0,
    eps: float = 0.0,
    output_dtype: int = 0,
) -> torch.Tensor:
    return split_embedding_codegen_lookup_function(
        host_weights, weights_offsets, D_offsets, total_D, indices, offsets, pooling_mode
    )


def expand_into_jagged_permute(
    permute: torch.Tensor, input_offsets: torch.Tensor, output_offsets: torch.Tensor, output_size: int
) -> torch.Tensor:
    return _hpex_C.expand_into_jagged_permute(permute, input_offsets, output_offsets, output_size)


def bounds_check_indices(
    rows_per_table: torch.Tensor,
    indices: torch.Tensor,
    offsets: torch.Tensor,
    bounds_check_mode: int,
    warning: torch.Tensor,
    weights: torch.Tensor | None,
):
    _hpex_C.bounds_check_indices(indices, offsets, warning, rows_per_table, bounds_check_mode, weights)


def split_permute_cat(
    input: torch.Tensor, indices: torch.Tensor, batch_size: int, num_features: int, dims: int
) -> torch.Tensor:
    return _hpex_C.split_permute_cat(input, indices, batch_size, num_features, dims)
